github/codeowners.py

```diff
@@ -116,38 +118,146 @@ def resolve_team_members(
     if not team_or_none:
         logger.warning('failed to lookup team {t}'.format(t=team_name))
         return []
-    for member in map(github_api.user, team_or_none.members()):
-        if member.email:
-            yield member.email
+    yield from map(github_api.user, team_or_none.members())
+
+
+@dataclasses.dataclass
+class CodeownerEntry:
+    value: str
+
+
+@dataclasses.dataclass
+class UsernameCodeownerEntry(CodeownerEntry):
+    pass
+
+
+@dataclasses.dataclass
+class EmailCodeownerEntry(CodeownerEntry):
+    pass
+
+
+@dataclasses.dataclass
+class TeamnameCodeownerEntry(CodeownerEntry):
+    pass
+
+
+class Filter:
+    @staticmethod
+    def usernames(codeowner_entry: CodeownerEntry) -> bool:
+        return isinstance(codeowner_entry, UsernameCodeownerEntry)
+
+    @staticmethod
+    def emails(codeowner_entry: CodeownerEntry) -> bool:
+        return isinstance(codeowner_entry, EmailCodeownerEntry)
+
+    @staticmethod
+    def teamnames(codeowner_entry: CodeownerEntry) -> bool:
+        return isinstance(codeowner_entry, TeamnameCodeownerEntry)
+
+
+def iter_codeowners(
+    codeowners_entries,
+    node_filter: typing.Callable[[CodeownerEntry], bool]=None,
+) -> typing.Generator[str, None, None]:
```

Comment:
I think the generator yields CodeownerEntry-instances, does it not? we might also omit inheritance and enumerate the three different types - it is not that many.